#!/usr/bin/env cwl-runner
class: CommandLineTool
cwlVersion: v1.0
label: 'GATK4: GetFilterMutectCalls'
doc: |-
  Filter variants in a Mutect2 VCF callset.

  FilterMutectCalls applies filters to the raw output of Mutect2. Parameters are contained in M2FiltersArgumentCollection and described in https://github.com/broadinstitute/gatk/tree/master/docs/mutect/mutect.pdf. To filter based on sequence context artifacts, specify the --orientation-bias-artifact-priors [artifact priors tar.gz file] argument one or more times. This input is generated by LearnReadOrientationModel.

  If given a --contamination-table file, e.g. results from CalculateContamination, the tool will additionally filter on contamination fractions. This argument may be specified with a table for one or more tumor sample. Alternatively, provide a numerical fraction to filter with the --contamination argument. FilterMutectCalls can also be given one or more --tumor-segmentation files, which are also output by CalculateContamination.

requirements:
- class: ShellCommandRequirement
- class: InlineJavascriptRequirement
- class: DockerRequirement
  dockerPull: broadinstitute/gatk:4.1.3.0

inputs:
- id: javaOptions
  label: javaOptions
  type:
  - type: array
    items: string
  - 'null'
- id: compression_level
  label: compression_level
  doc: |-
    Compression level for all compressed files created (e.g. BAM and VCF). Default value: 2.
  type:
  - int
  - 'null'
- id: contaminationTable
  label: contaminationTable
  doc: Tables containing contamination information.
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --contamination-table
- id: segmentationFile
  label: segmentationFile
  doc: |-
    Tables containing tumor segments' minor allele fractions for germline hets emitted by CalculateContamination
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --tumor-segmentation
- id: statsFile
  label: statsFile
  doc: The Mutect stats file output by Mutect2
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --stats
- id: readOrientationModel
  label: readOrientationModel
  doc: |-
    One or more .tar.gz files containing tables of prior artifact probabilities for the read orientation filter model, one table per tumor sample
  type:
  - File
  - 'null'
  inputBinding:
    prefix: --orientation-bias-artifact-priors
- id: vcf
  label: vcf
  doc: vcf to be filtered
  type: File
  secondaryFiles:
  - .tbi
  inputBinding:
    prefix: -V
- id: reference
  label: reference
  doc: Reference sequence file
  type: File
  secondaryFiles:
  - .fai
  - .amb
  - .ann
  - .bwt
  - .pac
  - .sa
  - ^.dict
  inputBinding:
    prefix: -R
- id: outputFilename
  label: outputFilename
  type:
  - string
  - 'null'
  default: generated.vcf.gz
  inputBinding:
    prefix: -O
    position: 2

outputs:
- id: out
  label: out
  doc: vcf containing filtered calls
  type: File
  secondaryFiles:
  - .tbi
  outputBinding:
    glob: generated.vcf.gz
    loadContents: false
stdout: _stdout
stderr: _stderr

baseCommand:
- gatk
- FilterMutectCalls
arguments:
- prefix: --java-options
  position: -1
  valueFrom: |-
    $("-Xmx{memory}G {compression} {otherargs}".replace(/\{memory\}/g, (([inputs.runtime_memory, 16, 4].filter(function (inner) { return inner != null })[0] * 3) / 4)).replace(/\{compression\}/g, (inputs.compression_level != null) ? ("-Dsamjdk.compress_level=" + inputs.compression_level) : "").replace(/\{otherargs\}/g, [inputs.javaOptions, []].filter(function (inner) { return inner != null })[0].join(" ")))
id: Gatk4FilterMutectCalls
