
from abc import ABC, abstractmethod

from janis_core import (
    ToolInput,
    Filename,
    Boolean,
    ToolOutput,
    InputSelector,
    ToolMetadata
)
from janis_core.tool.test_classes import TTestCase
from ..types import Vcf, CompressedVcf
from .bioinformaticstool import BioinformaticsTool


class VcfToolsToolBase(BioinformaticsTool, ABC):
    def tool_provider(self):
        return "VCFtools"

    @classmethod
    @abstractmethod
    def vcftools_command(cls):
        raise Exception("Tool or command not found")

    @classmethod
    def base_command(cls):
        return ["", cls.vcftools_command()]

    def inputs(self):
        return []

    def doc(self):
        return """vcftools is a suite of functions for use on genetic variation data in the form of VCF and BCF files. The tools provided will be used mainly to summarize data, run calculations on data, filter out data, and convert data into other useful file formats.
        
        Documentation:https://vcftools.github.io/man_latest.html#NAME""".strip()

    @abstractmethod
    def container(self):
        raise Exception(
            "An error likely occurred when resolving the method order for docker"
        )

    def arguments(self):
        return []



class VcfToolsvcftoolsBase(VcfToolsToolBase, ABC):
    def tool(self):
        return "VcfTools"

    @classmethod
    def vcftools_command(cls):
        return "vcftools"

    def inputs(self):
        return [
            # bcf is not supported yet
            ToolInput(
                "vcf",
                Vcf(),
                prefix="--vcf",
                doc="This option defines the VCF file to be processed. VCFtools expects files in VCF format v4.0, v4.1 or v4.2. The latter two are supported  with  some  small limitations. If the user provides a dash character '-' as a file name, the program expects a VCF file to be piped in through standard in.",
            ),
            ToolInput(
                "outputFilename",
                Filename(prefix=InputSelector("vcf", remove_file_extension=True)),
                prefix="--out",
                doc='<output_prefix>. This option defines the output filename prefix for all files generated by vcftools. For example, if <prefix> is set to output_filename, then all output files will be of the form output_filename.*** . If this option is omitted, all output files will have the prefix "out." in the current working directory.',
            ),
            *self.additional_inputs,
        ]

    def outputs(self):
        return [
            ToolOutput("out", Vcf, glob=InputSelector("outputFilename") + ".recode.vcf")
        ]

    def friendly_name(self):
        return "VcfTools"

    def bind_metadata(self):
        from datetime import date

        return ToolMetadata(
            contributors=["Jiaan Yu"],
            dateCreated=date(2020, 5, 22),
            dateUpdated=date(2020, 5, 22),
            institution="VCFtools",
            doi=None,
            citation=None,
            keywords=["vcftools"],
            documentationUrl="https://vcftools.github.io/man_latest.html",
            documentation="""NAME
vcftools v0.1.16 âˆ’ Utilities for the variant call format (VCF) and binary variant call format (BCF)

SYNOPSIS
vcftools [ --vcf FILE | --gzvcf FILE | --bcf FILE] [ --out OUTPUT PREFIX ] [ FILTERING OPTIONS ] [ OUTPUT OPTIONS ]

DESCRIPTION
vcftools is a suite of functions for use on genetic variation data in the form of VCF and BCF files. The tools provided will be used mainly to summarize data, run calculations on data, filter out data, and convert data into other useful file formats.
""".strip(),
        )

    additional_inputs = [
        ToolInput(
            "removeFileteredAll",
            Boolean(optional=True),
            prefix="--remove-filtered-all",
            position=1,
            shell_quote=False,
            doc="Removes all sites with a FILTER flag other than PASS.",
        ),
        ToolInput(
            "recode",
            Boolean(optional=True),
            prefix="--recode",
            position=1,
            shell_quote=False,
            doc="",
        ),
        ToolInput(
            "recodeINFOAll",
            Boolean(optional=True),
            prefix="--recode-INFO-all",
            position=1,
            shell_quote=False,
            doc="These options can be used with the above recode options to define an INFO key name to keep in the output  file.  This  option can be used multiple times to keep more of the INFO fields. The second option is used to keep all INFO values in the original file.",
        )
        # tbc
    ]

    def tests(self):
        parent_dir = "https://swift.rc.nectar.org.au/v1/AUTH_4df6e734a509497692be237549bbe9af/janis-test-data/bioinformatics"
        somatic_data = f"{parent_dir}/wgssomatic_data"
        return [
            TTestCase(
                name="basic",
                input={
                    "vcf": f"{somatic_data}/stdout.norm.vcf",
                    "removeFileteredAll": True,
                    "recode": True,
                    "recodeINFOAll": True,
                },
                output=Vcf.basic_test(
                    "out",
                    34393,
                    147,
                    ["GATKCommandLine"],
                    "c083775bc8c49397fb65ec12cd435688",
                ),
            ),
        ]


class VcfTools_0_1_16:
    def container(self):
        return "biocontainers/vcftools:v0.1.16-1-deb_cv1"

    def version(self):
        return "0.1.16"


class VcfToolsvcftools_0_1_16(VcfTools_0_1_16, VcfToolsvcftoolsBase):
    pass


VcfToolsvcftoolsLatest = VcfToolsvcftools_0_1_16
