#!/usr/bin/env cwltool

cwlVersion: v1.2 
class: CommandLineTool

baseCommand: [ /unlock/infrastructure/binaries/guppy/ont-guppy-cpu_v6.2.1/bin/guppy_basecaller ]

label: Basecalling of raw reads produced with ONT sequencing platforms like MinION and PromethION.
doc: |
  The workflow basecalles raw reads stored in the *.fast5 files created by the ONT sequencer with Guppy v6.0.1 for CPU.
  Original installation instructions require login credentials for the Nanopore Community.
  
  Direct download from:
    CPU: https://mirror.oxfordnanoportal.com/software/analysis/ont-guppy-cpu_6.2.1_linux64.tar.gz
    GPU: https://mirror.oxfordnanoportal.com/software/analysis/ont-guppy_6.2.1_linux64.tar.gz

requirements:
  - class: InlineJavascriptRequirement
  - class: InitialWorkDirRequirement
    listing:
    - entry: "$({class: 'Directory', listing: []})"
      entryname: "guppy_output"
      writable: true
    - entryname: nanopore_reads.txt
      entry: |-
        ${
          // let content = "";
          // if (inputs.fast5_files == null)
          //   return content;
          // for (var i = 0; i < inputs.fast5_files.length; i++) {
          //   content += inputs.fast5_files[i].path + "\n"
          // }
          // return content;
          let content = "";
          if (inputs.fast5_files == null)
            return content;
          for (var i = 0; i < inputs.fast5_files.length; i++) {
            content += inputs.fast5_files[i] + "\n"
          }
          return content;
        }

arguments:
  - valueFrom: "guppy_output" #output directory
    prefix: --save_path

inputs:
  fast5_directory:
    label: Directory with raw reads in *.fast5 format
    type: string?
    inputBinding:
      prefix: --input_path
  fast5_files:
    label: List of files in *.fast5 format
    type: string[]?
    inputBinding:
      prefix: --input_file_list 
      valueFrom: 'nanopore_reads.txt'
  threads:
    label: Number of basecallers to be run in parallel, each one with one thread
    type: int?
    default: 1
    inputBinding:
      prefix: --cpu_threads_per_caller
  configuration_command:
    label: Guppy preset with settings for the known flowcell and kit used for sequencing
    type: string
    inputBinding:
      prefix: --config
  # configuration_file: #optional
  #   label: Custom configuration file with defined settings
  #   type: File?
  #   inputBinding:
  #     prefix: --config
  # flowcell_id: #optional. Specify instead of configuration file. See: guppy_basecaller --print_workflows
  #   label: Identifier of flowcell used during ONT sequencing (e.g. FLO-MIN106)
  #   type: string?
  #   inputBinding:
  #     prefix: --flowcell
  # kit_id: #optional. Specify instead of configuration file. See: guppy_basecaller --print_workflows
  #   label: Identifier of the kit used during ONT sequencing (e.g. LSK109)
  #   type: string?
  #   inputBinding:
  #     prefix: --kit
  enable_adapter_trimming:
    label: Determines if Guppy performs adapter trimming
    type: boolean
    default: false
    inputBinding:
      prefix: --trim_barcodes
  recursive_input_search:
    label: Recursive search in subdirectories to find the *.fast5 files
    type: boolean
    default: true
    inputBinding:
      prefix: --recursive
  disable_qscore_filtering:
    label: disable qscore filtering
    type: boolean
    default: true
    inputBinding:
      prefix: --disable_qscore_filtering
  disable_pings:
    label: disable pings
    type: boolean
    default: true
    inputBinding:
      prefix: --disable_pings

outputs:
  reads_directory:
    label: Directory with reads in fastq format and summary generated by Guppy
    type: Directory
    outputBinding:
      glob: guppy_output
  fastq_reads:
    type: File[]
    outputBinding:
      glob: guppy_output/*.fastq
  summary:
    type: File
    outputBinding:
      glob: guppy_output/sequencing_summary.txt
  telemetry:
    type: File
    outputBinding:
      glob: guppy_output/sequencing_telemetry.js
  guppy_log:
    type: File[]
    outputBinding:
      glob: guppy_output/*.log

s:author:
  - class: s:Person
    s:identifier: https://orcid.org/0000-0002-5516-8391
    s:email: mailto:german.royvalgarcia@wur.nl
    s:name: GermÃ¡n Royval
  - class: s:Person
    s:identifier: https://orcid.org/0000-0001-8172-8981
    s:email: mailto:jasper.koehorst@wur.nl
    s:name: Jasper Koehorst
  - class: s:Person
    s:identifier: https://orcid.org/0000-0001-9524-5964
    s:email: mailto:bart.nijsse@wur.nl
    s:name: Bart Nijsse

s:citation: https://m-unlock.nl
s:codeRepository: https://gitlab.com/m-unlock/cwl
s:dateCreated: "2021-11-22"
s:license: https://spdx.org/licenses/Apache-2.0 
s:copyrightHolder: "UNLOCK - Unlocking Microbial Potential"

$namespaces:
  s: https://schema.org/
  